src/com/xilinx/rapidwright/rwroute/RWRoute.java

```diff
@@ -0,0 +1,1841 @@
+		
+		Node sourceINTNode = RouterHelper.projectOutputPinToINTNode(altSource);
+		Routable sourceR = this.createAddRoutableNode(this.rnodeId, altSource, sourceINTNode, RoutableType.PINFEED_O);;
+		for(Connection otherConnectionOfNet : netWrapper.getConnections()) {
+			otherConnectionOfNet.setSource(altSource);
+			otherConnectionOfNet.setSourceRnode(sourceR);
+		}
+			
+		return true;
+	}
+	
+	/**
+	 * Unroutes preserved nets to release routing resource to resolve congestion that blocks the routablity of a connection.
+	 * @param connection The connection in question.
+	 * @return The number of unrouted nets.
+	 */
+	private int unrouteReservedNetsToReleaseResources(Connection connection) {
+		// Find those reserved signals that are using uphill nodes of the target pin node
+		Set<Net> toRouteNets = new HashSet<>();
+		for(Node node : connection.getSinkRnode().getNode().getAllUphillNodes()) {
+			Net toRoute = this.preservedNodes.get(node);
+			if(toRoute == null) continue;
+			if(toRoute.isClockNet() || toRoute.isStaticNet()) continue;
+			toRouteNets.add(toRoute);
+		}
+		// Find those preserved nets that are using downhill nodes of the source pin node
+		for(Node node : connection.getSourceRnode().getNode().getAllDownhillNodes()) {
+			Net toRoute = this.preservedNodes.get(node);
+			if(toRoute == null) continue;
+			if(toRoute.isClockNet() || toRoute.isStaticNet()) continue;
+			toRouteNets.add(toRoute);
+		}
+		
+		if(!toRouteNets.isEmpty()) {
+			System.out.println("INFO: Unroute " + toRouteNets.size() + " preserved nets");
+			System.out.println(toRouteNets);
+		}
+		
+		for(Net n : toRouteNets) {
+			List<Node> reservedNetNodes = RouterHelper.getNodesOfNet(n);
+			for(Node toRemove : reservedNetNodes) {
+				this.preservedNodes.remove(toRemove);
+			}
+			
+			NetWrapper netnew = this.createsNetWrapperAndConnections(n, this.config.getBoundingBoxExtensionX(), this.config.getBoundingBoxExtensionY(), this.multiSLRDevice);
+			
+			for(int i = 0; i < reservedNetNodes.size(); i ++) {
+				Node toBuild = reservedNetNodes.get(i);
+				Routable rnode = this.rnodesCreated.get(toBuild);
+				if(rnode == null) {
+					rnode = new RoutableNode(this.rnodeId, toBuild, RoutableType.WIRE);
+					this.rnodeId++;
+					this.rnodesCreated.put(toBuild, rnode);
+				}
+				// Each rnode created above should be added to its parents if parent exists,
+				// because children of an existing parent may have been set.
+				for(Node uphill : toBuild.getAllUphillNodes()) {
+					// Without this routethru check, there will be Invalid Programming for Site error shown in Vivado.
+					// Do not use those nodes, because we do not know if the routethru is available or not
+					if(routethruHelper.isRouteThru(uphill, toBuild)) continue;
+					RoutableNode parent = (RoutableNode) this.rnodesCreated.get(uphill);
+					if(parent != null && !parent.childrenNotSet()) {
+						if(!parent.getChildren().contains(rnode)) parent.getChildren().add(rnode);
+					}
+				}
+			}
```

Comment:
Can this loop be merged with the one above?

Reply:
Merged.